### ISSUE TITLE
Add an option to conform (fully) to the Protobuf 3 canonical JSON representation

### ISSUE THREAD

**Issue by mindloaf** [HubSpot/jackson-datatype-protobuf#76]
int64, fixed64, uint64: JSON value will be a decimal string. Either numbers or strings are accepted.

If the Any contains a value that has a special JSON mapping, it will be converted as follows: `{"@type": xxx, "value": yyy}`. Otherwise, the value will be converted into a JSON object, and the `"@type"` field will be inserted to indicate the actual data type.

**Comment by jhaber** [comment]
Thanks for opening this issue, I think it would be a great improvement 👍

**Comment by jhaber** [comment]
btw I started implementing this on the `compatible-serialization` branch, but the changes kept growing in scope and I haven't found the time to finish them up. As part of that work I realized that we don't actually serialize/deserialize uint32/uint64/fixed32/fixed64 correctly (depending on size, we serialize them as negative numbers, rather than inspecting the sign bit and taking the correct action)

**Comment by jhaber** [comment]
And I think support for serializing/deserializing `DynamicMessage` support is going to be pretty unlikely (might as well just use `com.google.protobuf.util.JsonFormat` for that)

**Comment by mindloaf** [comment]
Yeah DynamicMessages are definitely more "nice to have". Consistent handling of the more basic types is definitely more important.

**Comment by mindloaf** [comment]
This recently bit me as well:
> Proto3 JSON parsers are required to accept both the converted lowerCamelCase name and the proto field name.
I ran into some JSON that was using both camelCased and snake_cased field names. I ended up having to hack in JsonFormat (essentially double parsing) to handle both.

**Comment by jhaber** [comment]
That should be handled with this option:
https://github.com/HubSpot/jackson-datatype-protobuf/blob/5d403ef47a07a13433cfdca08fb6e6fe1ec7a96b/src/main/java/com/hubspot/jackson/datatype/protobuf/ProtobufJacksonConfig.java#L41

Did you have a case where that wasn't sufficient?

**Comment by mindloaf** [comment]
To be clear, this was a case were the field names were both literal (snake_cased, in this instance) and camelCased (not literal).

**Comment by mindloaf** [comment]
I'll try out `acceptLiteralFieldnames`, but I think we also had an issue related to https://github.com/HubSpot/jackson-datatype-protobuf/issues/81

**Comment by jhaber** [comment]
There's an option on `ProtobufJacksonConfig.Builder` now has a `useCanonicalSerialization` method. There are still some gaps (I believe these are mostly related to the handling of `Any`), but using that method will automatically pick up any fixes/features in the future
